File: src/seurat/bin/transform/sc_normalization.R
#!/usr/bin/env Rscript

library("argparse")
suppressMessages(library("Seurat", quietly = TRUE))

parser <- ArgumentParser(description = "Normalize and scale data using SCT")

parser$add_argument(
    "--input",
    type = "character",
    dest = "input",
    action = "store",
    help = "A Rds file containing a Seurat object."
)
parser$add_argument(
    "--output",
    type = "character",
    dest = "output",
    action = "store",
    help = "Output filename."
)
parser$add_argument(
    "--method",
    type = "character",
    dest = "method",
    action = "store",
    default = "LogNormalize",
    help = "Method for normalization. Choose one of: LogNormalize, CLR, RC"
)
parser$add_argument(
    "--scale-factor",
    type = "integer",
    dest = "scale_factor",
    action = "store",
    default = 10000,
    help = "Sets the scale factor for cell-level normalization."
)

args <- parser$parse_args()
print(args)

seuratObj <- tryCatch({
    readRDS(file = args$input)
}, warning = function(w) {
}, error = function(e) {
    stop("VSN ERROR: Cannot read the given Rds file.")
}, finally = {})

if (class(x = seuratObj) != "Seurat") {
    stop("VSN ERROR: The object contained in the Rds file is not a Seurat object.")
}

seuratObj <- NormalizeData(
    object = seuratObj,
    normalization.method = args$method,
    scale.factor = args$scale_factor
)

saveRDS(
    object = seuratObj,
    file = args$output
)

File: src/scanpy/bin/transform/sc_normalization.py
#!/usr/bin/env python3

import argparse
import os
import scanpy as sc

parser = argparse.ArgumentParser(description='')

parser.add_argument(
    "input",
    type=argparse.FileType('r'),
    help='Input h5ad file.'
)

parser.add_argument(
    "output",
    type=argparse.FileType('w'),
    help='Output h5ad file.'
)

parser.add_argument(
    "-x", "--method",
    type=str,
    action="store",
    dest="method",
    default="cpx",
    help="Normalize the data. Choose one of : cpx"
)

parser.add_argument(
    "-f", "--target-sum",
    type=int,
    action="store",
    dest="target_sum",
    default=1e4,
    help="Multiplying factor used when running 'cpx' method. If None, after normalization, each observation (cell) has a total count equal to the median of total counts for observations (cells) before normalization."
)

args = parser.parse_args()

# Define the arguments properly
FILE_PATH_IN = args.input
FILE_PATH_OUT_BASENAME = os.path.splitext(args.output.name)[0]

# I/O
# Expects h5ad file
try:
    adata = sc.read_h5ad(filename=FILE_PATH_IN.name)
except IOError:
    raise Exception("VSN ERROR: Wrong input format. Expects .h5ad files, got .{}".format(os.path.splitext(FILE_PATH_IN)[0]))

#
# Normalize the data
#

adata.raw = adata

if args.method == 'cpx':
    # Total-count normalize (library-size correct) to '-r' reads/cell
    sc.pp.normalize_total(
        adata,
        target_sum=args.target_sum
    )
else:
    raise Exception("VSN ERROR: Method does not exist.")

# I/O
adata.write_h5ad("{}.h5ad".format(FILE_PATH_OUT_BASENAME))
